src/main/java/net/revelc/code/apilyzer/maven/plugin/AnalyzeMojo.java

```diff
@@ -264,32 +339,145 @@ public URL apply(String input) {
+      } else {
+        annotations = new Annotation[0];
+      }
+
+
+      for (String pattern : includeAnnotations) {
+        for (Annotation annotation : annotations) {
+          if (annotation.toString().matches(pattern)) {
+            if (!annotationExcludes(annotations)) {
+              addPublicApiType(publicApiClasses, publicSet, classInfo);
             }
+            continue classLoop;
           }
-          if (!exclude) {
-            Class<?> clazz = classInfo.load();
-            if (isPublicOrProtected(clazz)) {
-              publicApiClasses.add(clazz);
-              publicSet.add(clazz.getName());
-            }
+        }
+      }
+
+
+      for (String pattern : includes) {
+        if (classInfo.getName().matches(pattern)) {
+          if (!annotationExcludes(annotations)) {
+            addPublicApiType(publicApiClasses, publicSet, classInfo);
           }
           break;
         }
+
+      }
+    }
+  }
+
+  private void addPublicApiType(List<Class<?>> publicApiClasses, TreeSet<String> publicSet,
+      ClassInfo classInfo) {
+    Class<?> clazz = classInfo.load();
+    if (isPublicOrProtected(clazz)) {
+      publicApiClasses.add(clazz);
+      publicSet.add(clazz.getName());
+
+      // for top level classes, add their inner classes
+      if (clazz.getEnclosingClass() == null) {
+        addPublicInnerClasses(publicApiClasses, publicSet, clazz);
+      }
+    }
+  }
+
+  private boolean patternExcludes(ClassInfo classInfo) {
+    for (String excludePattern : excludes) {
+      if (classInfo.getName().matches(excludePattern)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  private boolean patternExcludes(Class<?> clazz) {
+    for (String excludePattern : excludes) {
+      if (clazz.getName().matches(excludePattern)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  private boolean annotationExcludes(Annotation[] annotations) {
+    for (String pattern : excludeAnnotations) {
+      for (Annotation annotation : annotations) {
+        if (annotation.toString().matches(pattern)) {
+          return true;
+        }
+      }
+    }
+
+    return false;
+  }
+
+  private boolean annotationExcludes(Class<?> clazz) {
+    if (excludeAnnotations.size() == 0) {
+      return false;
+    }
+
+    Annotation[] annotations = getAnnotations(clazz);
+
+    for (String pattern : excludeAnnotations) {
+      for (Annotation annotation : annotations) {
+        if (annotation.toString().matches(pattern)) {
+          return true;
+        }
+      }
+    }
+
+    return false;
+  }
+
+  private void addPublicInnerClasses(List<Class<?>> publicApiClasses, TreeSet<String> publicSet,
+      Class<?> clazz) {
+    Class<?>[] innerClasses = clazz.getDeclaredClasses();
+    for (Class<?> ic : innerClasses) {
+      // If a class is in the Public API then all of its public inner class are also considered
+      // to be in the public API unless explicitly excluded.
```

Comment:
Why would this comment be true? Shouldn't it depend on whether the inner class matches the include pattern?

Reply:
No, the way it behaved before this change was that is a type matched a include then everything public (methods, fields, inner classes, etc) in that type was considered public API.  So I refined the behavior to allow excludes to be able to apply to these inner classes.